cpp/src/arrow/array/builder_dict.cc

```diff
@@ -145,229 +144,176 @@ Status DictionaryType::Unify(MemoryPool* pool, const std::vector<const DataType*
-    const std::shared_ptr<DataType>& type, MemoryPool* pool)
-    : ArrayBuilder(type, pool),
-      memo_table_(new MemoTableImpl(0)),
-      delta_offset_(0),
-      byte_width_(checked_cast<const FixedSizeBinaryType&>(*type).byte_width()) {}
-
-template <typename T>
-void DictionaryBuilder<T>::Reset() {
-  ArrayBuilder::Reset();
-  values_builder_.Reset();
-  memo_table_.reset(new MemoTableImpl(0));
-  delta_offset_ = 0;
-}
+    template <typename DataType, typename Array>
+    Status InsertValues(
+        const DataType&, const Array& array,
+        typename internal::DictionaryTraits<DataType>::MemoTableType* = nullptr) {
+      for (int64_t i = 0; i < array.length(); ++i) {
+        ARROW_IGNORE_EXPR(impl_->GetOrInsert(array.GetView(i)));
+      }
+      return Status::OK();
+    }
+  };
+
+  struct ArrayDataGetter {
+    std::shared_ptr<DataType> value_type_;
+    MemoTable* memo_table_;
+    MemoryPool* pool_;
+    int64_t start_offset_;
+    std::shared_ptr<ArrayData>* out_;
+
+    Status Visit(const DataType&, void* = nullptr) {
+      return Status::NotImplemented("Getting array data of ", value_type_,
+                                    " is not implemented");
+    }
 
-template <typename T>
-Status DictionaryBuilder<T>::Resize(int64_t capacity) {
-  RETURN_NOT_OK(CheckCapacity(capacity, capacity_));
-  capacity = std::max(capacity, kMinBuilderCapacity);
+    template <typename T>
+    Status Visit(const T&,
+                 typename internal::DictionaryTraits<T>::MemoTableType* = nullptr) {
+      using ConcreteMemoTable = typename internal::DictionaryTraits<T>::MemoTableType;
+      auto memo_table = static_cast<ConcreteMemoTable*>(memo_table_);
+      return internal::DictionaryTraits<T>::GetDictionaryArrayData(
+          pool_, value_type_, *memo_table, start_offset_, out_);
+    }
+  };
 
-  if (capacity_ == 0) {
-    // Initialize hash table
-    // XXX should we let the user pass additional size heuristics?
-    delta_offset_ = 0;
+ public:
+  explicit DictionaryMemoTableImpl(const std::shared_ptr<DataType>& type)
+      : type_(type), memo_table_(nullptr) {
+    MemoTableInitializer visitor{type_, &memo_table_};
+    ARROW_IGNORE_EXPR(VisitTypeInline(*type_, &visitor));
   }
-  RETURN_NOT_OK(values_builder_.Resize(capacity));
-  capacity_ = values_builder_.capacity();
-  return Status::OK();
-}
-
-Status DictionaryBuilder<NullType>::Resize(int64_t capacity) {
-  RETURN_NOT_OK(CheckCapacity(capacity, capacity_));
-  capacity = std::max(capacity, kMinBuilderCapacity);
 
-  RETURN_NOT_OK(values_builder_.Resize(capacity));
-  capacity_ = values_builder_.capacity();
-  return Status::OK();
-}
+  Status InsertValues(const std::shared_ptr<Array>& array) {
+    ArrayValuesInserter visitor{this};
+    return VisitArrayInline(*array, &visitor);
+  }
 
-template <typename T>
-Status DictionaryBuilder<T>::Append(const Scalar& value) {
-  RETURN_NOT_OK(Reserve(1));
+  template <typename T>
```

Comment:
This seems roundabout; for integer values you could use `ConcreteMemoTable = typename internal::HashTraits<typename CTypeTraits<T>::ArrowType>` and for GetOrInsert(string_view) you could just explicitly name `BinaryMemoTable`

Reply:
Thanks! Done.
I didn't know that we have `CTypeTraits`!